### PR TITLE
release: Copy oscontainer to quay.io

### PR DESCRIPTION
**Pull request by cheesesashimi**
This PR is intended to supercede and incorporates elements of: https://github.com/coreos/fedora-coreos-pipeline/pull/383

### CODE REVIEW

File: manifests/pod.yaml [hunk 1]
```diff
@@ -54,6 +54,9 @@ spec:
      - name: fedora-messaging-coreos-key
        mountPath: /run/kubernetes/secrets/fedora-messaging-coreos-key
        readOnly: true
+     - name: oscontainer-registry
```
Comment (jlebon) [inline manifests/pod.yaml:57]: The pod shouldn't need the secret mounted in since we're using `withCredentials`.

File: Jenkinsfile.release [hunk 2]
```diff
@@ -85,9 +85,21 @@ podTemplate(cloud: 'openshift', label: pod_label, yaml: pod) {
             cosa init --branch ${params.STREAM} https://github.com/coreos/fedora-coreos-config
             cosa buildfetch --build=${params.VERSION} \
                 --arch=all --url=s3://${s3_stream_dir}/builds
+            cosa buildprep --ostree --build=${params.VERSION} \
```
Comment (jlebon) [inline Jenkinsfile.release:88]: You can add `--artifact ostree` to the previous `buildfetch` call instead.

File: Jenkinsfile.release [hunk 3]
```diff
@@ -85,9 +85,21 @@ podTemplate(cloud: 'openshift', label: pod_label, yaml: pod) {
...
+                --arch=all s3://${s3_stream_dir}/builds
             """)
         }
 
+        stage('Sync oscontainer to quay.io') {
+            withCredentials([string(credentialsId: "oscontainer-secret", variable: "OSCONTAINER_SECRET")]) {
+                shwrap("""ociarchive=\$(cosa meta --image-path ostree)
+                        case \${ociarchive} in
+                        *ociarchive) skopeo copy --authfile=<($OSCONTAINER_SECRET) oci-archive://\${ociarchive} docker://quay.io/coreos-assembler/fcos:${params.STREAM};;
```
Comment (jlebon) [inline Jenkinsfile.release:97]: Might be good to const-ify the quay.io URL here at the beginning of the file for clarity. And also a comment there that says long-term we want `quay.io/fedora/coreos` as per https://fedoraproject.org/wiki/Changes/OstreeNativeContainer.

File: Jenkinsfile.release [hunk 4]
```diff
@@ -85,9 +85,21 @@ podTemplate(cloud: 'openshift', label: pod_label, yaml: pod) {
...
                 --arch=all --url=s3://${s3_stream_dir}/builds
+            cosa buildprep --ostree --build=${params.VERSION} \
+                --arch=all s3://${s3_stream_dir}/builds
             """)
         }
 
+        stage('Sync oscontainer to quay.io') {
+            withCredentials([string(credentialsId: "oscontainer-secret", variable: "OSCONTAINER_SECRET")]) {
+                shwrap("""ociarchive=\$(cosa meta --image-path ostree)
+                        case \${ociarchive} in
```
Comment (jlebon) [inline Jenkinsfile.release:96]: All streams use ociarchive now and at some point we'll probably rip out the old support for .tar in cosa. So I think we can just simplify this logic here.

File: jenkins/config/oscontainer-secret.yaml [hunk 5]
```diff
@@ -0,0 +1,9 @@
+credentials:
+  system:
+    domainCredentials:
+      - credentials:
+        - string:
```
Comment (jlebon) [inline jenkins/config/oscontainer-secret.yaml:5]: Cool with this, though feels like a secret file would be more appropriate for this instead.

Comment (jlebon) [reply jenkins/config/oscontainer-secret.yaml:5]: See https://jenkins-fedora-coreos.apps.ocp.ci.centos.org/configuration-as-code/reference#Credentials-file.

Comment (cheesesashimi) [reply jenkins/config/oscontainer-secret.yaml:5]: I was originally intending to do that, I just couldn't find the syntax for some reason. Thanks for the pointer!

Comment (cheesesashimi) [reply jenkins/config/oscontainer-secret.yaml]: It took a bit of trial-and-error. The short version is that you have to base64-encode the file contents to make it work.

File: Jenkinsfile.release (hunk 2 above)
Comment (cheesesashimi) [reply Jenkinsfile.release]: I'm not sure that works:

```console
$ cosa buildfetch --ostree
usage: cmd-buildfetch [-h] [--url URL] [--stream STREAM] [-b BUILD] [--force] [--arch ARCH] [--artifact ARTIFACT]
cmd-buildfetch: error: unrecognized arguments: --ostree
```

**Review by jlebon — APPROVED** [review]
Thanks, looks great!

Final bits:
- Can you squash your two commits into one?
- Can you add a section in HACKING.md about the new secret similar to the other ones?

**Comment by cheesesashimi** [comment]
Done!

File: HACKING.md [hunk 6]
```diff
@@ -218,6 +218,13 @@ uuidgen -r > secret
 oc create secret generic github-webhook-shared-secret --from-file=secret
 ```
 
+### [PROD] Create quay.io image push secret
+
+This secret is used to push the resulting OCI image to Quay.io
+
+1. Obtain the file `oscontainer-secret` from BitWarden.
+2. Run: `$ oc create secret generic oscontainer-secret --from-file=dockercfg=oscontainer-secret`.
```
Comment (jlebon) [inline HACKING.md:226]: OK, *really* final bits so we don't break the pipeline as soon as this merges:
1. Did you add the secret to BitWarden? I don't see it.
2. Did you already create the OpenShift secret? I don't see it either.

**Comment by jlebon** [comment]
This is ready for merge. Waiting until our current test runs in the new infra are over before I get it in and redeploy Jenkins.

File: HACKING.md (hunk 6 above)
Comment (jlebon) [reply HACKING.md:226]: Secret was added to BitWarden and OpenShift (both old and new cluster in case we rollback the migration temporarily).

**Comment by jlebon** [comment]
I totally missed this until now but: when I said to squash your commits I meant have a single commit on top of Colin's commit. Authorship recognition is important and you did some nice cleanups and testing for this. Sorry for not noticing earlier!

**Comment by cgwalters** [comment]
In this case the [Co-authored-by tag](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors) seems best but for the record I don't think this is a big deal from my PoV - just something to keep in mind elsewhere!

I particularly try to preserve credit in the case of a new contributor to a project submitting a PR, and then I need to changes on top.  But that's not this case really.

Anyways - cool to see this landed!

**Comment by cheesesashimi** [comment]
No worries! I thought about it almost immediately after I squashed it.

**Pull request merged**